airgun/entities/subscription.py

```diff
@@ -0,0 +1,180 @@
+
+from airgun.entities.base import BaseEntity
+from airgun.navigation import NavigateStep, navigator
+from airgun.views.subscription import (
+    SubscriptionListView,
+    ManageManifestView,
+    DeleteManifestConfirmationView,
+    AddSubscriptionView,
+    SubscriptionDetailsView,
+)
+
+
+class SubscriptionEntity(BaseEntity):
+    def _wait_for_process_to_finish(self, has_manifest=False):
+        view = self.navigate_to(self, 'All')
+        view.progressbar.wait_displayed(timeout=20)
+        wait_for(
+                lambda: not view.progressbar.is_displayed,
+                timeout=60*10,
+                logger=view.progressbar.logger
+        )
+        wait_for(
+                lambda: self.has_manifest == has_manifest,
+                timeout=10,
+                logger=view.progressbar.logger
+        )
+        view.flash.dismiss()
+
+    @property
+    def has_manifest(self):
+        view = self.navigate_to(self, 'All')
+        return "disabled" not in view.browser.classes(view.add_button)
+
+    def add_manifest(self, manifest_file):
+        view = self.navigate_to(self, 'Manage Manifest')
+        view.fill({
+            'manifest.manifest_file': manifest_file,
+        })
+        self._wait_for_process_to_finish(has_manifest=True)
+
+    def delete_manifest(self, really=True):
+        view = self.navigate_to(self, 'Delete Manifest Confirmation')
+        if not really:
+            view.cancel_button.click()
+            try:
+                while view.is_displayed:
+                    pass
+            except NoSuchElementException:
+                pass
+            self.navigate_to(self, 'Manage Manifest').close_button.click()
+            return
+
+        view.delete_button.click()
+        self._wait_for_process_to_finish(has_manifest=False)
+
+    def read_delete_manifest_message(self):
+        view = self.navigate_to(self, 'Delete Manifest Confirmation')
+        return view.message.read()
+
+    def add(self, entity_name, quantity=1):
+        view = self.navigate_to(self, 'Add')
+        for row in view.table.rows(subscription_name=entity_name):
+            row['Quantity to Allocate'].fill(quantity)
+        view.submit_button.click()
+        self._wait_for_process_to_finish(has_manifest=True)
+
+    def search(self, value):
+        view = self.navigate_to(self, 'All')
+        return view.search(value)
+
+    def provided_products(self, entity_name):
+        view = self.navigate_to(self, 'Details', subscription=entity_name)
+        return [elem.text
+                for elem
+                in view.details.provided_products.browser.elements("./li")]
+
+    def enabled_products(self, entity_name):
+        locator = ("//div[contains(@class, 'list-group')]"
+                   "//div[contains(@class, 'list-group-item-heading')]"
+                   )
+        view = self.navigate_to(self, 'Details', subscription=entity_name)
+        view.enabled_products.select()
+        try:
```

Comment:
any specific reason to have additional exception handling here? `NoSuchElementException` sounds pretty good

Reply:
Yes, I want Airgun method to always return list - even if it is empty list. Methods like `search` already return empty list when no elements are found and I wanted to be consistent.